cds_ils/importer/eitems/importer.py

```diff
@@ -0,0 +1,116 @@
+
+        if not existing_provider:
+            return False
+
+        existing_priority = CDS_ILS_IMPORTER_PROVIDERS[
+            existing_provider]['priority']
+
+        # existing_priority = 0, self.priority = 1, returns False
+        return existing_priority >= self.current_provider_priority
+
+    def update_eitems(self, matched_document):
+        """Update eitems for given document."""
+        import ipdb; ipdb.set_trace()
+        document_pid = matched_document["pid"]
+
+        results = get_eitems_for_document_by_provider(document_pid,
+                                                      self.metadata_provider)
+
+        if len(results) == 0:
+            self.import_eitem(matched_document)
+        elif len(results) == 1:
+            existing_eitem = EItem.get_record_by_pid(results.hits[0].pid)
+            metadata_to_update = {}
+            self._build_eitem_dict(
+                {}, matched_document["pid"])
+            existing_eitem.update(metadata_to_update)
+            existing_eitem.commit()
+        else:
+            for hit in results:
+                existing_eitem = EItem.get_record_by_pid(hit.pid)
+                existing_eitem.delete()
+            self.import_eitem(matched_document)
+
+        if self.delete_lower_priority_providers:
+            document_eitems = EItemSearch().search_by_document_pid(
+                matched_document["pid"])
+            for hit in document_eitems:
+                eitem = EItem.get_record_by_pid(hit.pid)
+                if self._should_overwrite_eitem(eitem):
+                    eitem.delete()
+
+        db.session.commit()
+
+    def _build_eitem_dict(self, eitem_json, document_pid):
+        self._apply_url_login(eitem_json)
+        self._set_record_import_source(eitem_json)
+        dois = [doi for doi in self.json_data.get('identifiers', []) if
+                doi['scheme'] == 'DOI']
+        eitem_json.update({"document_pid": document_pid,
+                           "open_access": self.open_access,
+                           "identifiers": dois,
+                           "created_by": {"type": "import",
+                                          "value": self.metadata_provider}
+                           })
+
+    def _apply_url_login(self, eitem):
+        if 'urls' in eitem:
+            for url in eitem['urls']:
+                url['login_required'] = self.login_required
+
+    def import_eitem(self, new_document):
+        """Update eitems for given document."""
+
+        eitem_json = self.json_data.get('_eitem', None)
+
+        if eitem_json:
+            self._build_eitem_dict(eitem_json, new_document["pid"])
+            record_uuid = uuid.uuid4()
+            provider = EItemIdProvider.create(
+                object_type="rec",
+                object_uuid=record_uuid,
+            )
+
+            eitem_json["pid"] = provider.pid.pid_value
+            eitem = EItem.create(eitem_json, record_uuid)
```

Comment:
can it fail validation? Better to rollback just in case? See my comment above.

Reply:
added the try/except block